packages/vite/src/node/plugins/clientInjections.ts

```diff
@@ -49,12 +49,6 @@ export function clientInjectionsPlugin(config: ResolvedConfig): Plugin {
           .replace(`__HMR_PORT__`, JSON.stringify(port))
           .replace(`__HMR_TIMEOUT__`, JSON.stringify(timeout))
           .replace(`__HMR_ENABLE_OVERLAY__`, JSON.stringify(overlay))
-      } else if (code.includes('process.env.NODE_ENV')) {
-        // replace process.env.NODE_ENV
-        return code.replace(
-          /\bprocess\.env\.NODE_ENV\b/g,
-          JSON.stringify(config.mode)
-        )
       }
```

Comment:
@bluwy I saw you did a PR before touching NODE_ENV replacement. Do you know why this branch is needed? Looks to me (as suggested by @sheremet-va) that the replacement done in the DefinePlugin should already take care of `process.env.NODE_ENV`. Why is this also being replaced here?

Reply:
I ran into this a couple of days ago when I wanted to configure some env variables for a service worker. It turned out that `DefinePlugin` only takes place for the production build and the server build (when SSR is enabled). As for the development build, it is `vite/client.js` that is supposed to define env variables at runtime.

https://github.com/vitejs/vite/blob/3f27d58036bfe4149e563a105c08c2d4771db288/packages/vite/src/node/plugins/define.ts#L84-L89

The problem is, I would've expected that importing `vite/client.js` would define `process.env.NODE_ENV`. However, the reality is that it gets replaced by this workaround.

So I would be interested in listening to why it was decided to have this workaround instead of defining all the variables in `vite/client.js` at runtime (in dev mode).

Reply:
Yeah this look superfluous. The define plugin currently doesn't handle this case though since it bails out and relied on a global `globalThis.process.env.NODE_ENV = "env"` call for performance.

https://github.com/vitejs/vite/blob/56ed808360b1374b468745c5a8727ed9763924f6/packages/vite/src/node/plugins/define.ts#L85-L88

But it looks like currently we don't do a `globalThis.process.env.NODE_ENV = "env"` call anywhere, which causes the CI to fail. It should be fixable though if we add `process.env.NODE_ENV` to the `__DEFINES__`:

https://github.com/vitejs/vite/blob/3f27d58036bfe4149e563a105c08c2d4771db288/packages/vite/src/node/plugins/clientInjections.ts#L46

The current code now looks like a quick way to get it working before, which causes the Vitest issue since it doesn't handle `preventAssignment` cases.

Reply:
So instead of having `processNodeEnv` in https://github.com/vitejs/vite/blob/main/packages/vite/src/node/plugins/define.ts#L10
we should move it to the the place where Vite resolves config?

Reply:
I'm thinking of directly adding it in client injection https://github.com/vitejs/vite/pull/6935/commits/fa7e458c189c873f1a0c8a5c2dcc76e70fe8401f